mesonbuild/dependencies/mpi.py

```diff
@@ -244,3 +247,48 @@ def __init__(self, name: str, env: 'Environment', kwargs: T.Dict[str, T.Any],
         self.compile_args = ['-I' + incdir, '-I' + os.path.join(incdir, post)]
         if self.language == 'fortran':
             self.link_args.append('-l' + os.path.join(libdir, 'msmpifec'))
+
+
+class IMPIDependency(SystemDependency):
+
+    """Intel(R) MPI for Windows."""
+
+    def __init__(self, name: str, env: Environment, kwargs: T.Dict[str, T.Any],
+                 language: T.Optional[str] = None):
+        super().__init__(name, env, kwargs, language=language)
+        # only for windows
+        if not self.env.machines[self.for_machine].is_windows():
+            return
+        # only for x86_64
+        if detect_cpu_family(self.env.coredata.compilers.host) != 'x86_64':
+            return
+
+        rootdir = os.environ.get('I_MPI_ROOT')
+        if rootdir is None:
+            self.is_found = False
+            return
+
+        incdir = os.path.join(rootdir, 'include')
+        libdir = os.path.join(rootdir, 'lib')
+
+        buildtype = env.coredata.optstore.get_value_for(OptionKey('buildtype'))
+        assert isinstance(buildtype, str)
+        libdir_post = 'debug' if buildtype.startswith('debug') else 'release'
+        for subdirs in (['mpi', libdir_post], [libdir_post]):
+            libdir_buildtype = os.path.join(libdir, *subdirs)
+            if os.path.isdir(libdir_buildtype):
+                libdir = libdir_buildtype
+                break
+
+        found_header = os.path.isfile(os.path.join(incdir, 'mpi.h'))
+        found_library = os.path.isfile(os.path.join(libdir, 'impi.lib'))
+        if not found_header or not found_library:
+            self.is_found = False
+            return
+
+        self.is_found = True
+        self.compile_args = ['-I' + incdir]
+        self.link_args = ['-l' + os.path.join(libdir, 'impi')]
+        if self.language == 'cpp':
+            if os.path.isfile(os.path.join(libdir, 'impicxx.lib')):
+                self.link_args = ['-l' + os.path.join(libdir, 'impicxx')]
```

Comment:
Is it valid for a c++ dependency to not find this library?

Reply:
Most C++ codes use MPI though the C API and library, and not the long ago deprecated C++ bindings. I'm adding the library if it is there just in case. Depending on how you install Intel MPI (standalone installer, nuget, Python wheel) the `impicxx.lib` import library may be present or not. Adding it to the link line will not harm. However, not finding MPI just because of this library IMHO will do more harm than good.